### PR TITLE
Development

### PR DESCRIPTION
**Pull request by yourgirlmjayy**
This pull request introduces several helper functions to determine recommended activities and goals for users based on their location, weather, and past activities, past moods as well as the day of the week. 

**The functions include:**
Fetch and process weather data:
fetchWeatherDataWithAPI: Fetches weather data from API for a given location and date
fetchWeatherDatafromDB: Fetches weather data from database for a given user

Analyze user activity and mood patterns:
getDayOfWeek: Returns the day of the week for a given date
getTimeOfDay: Returns the time of day (morning, afternoon, evening) for a given date
getMonthOfYear: Returns the month of the year for a given date
isNegativeActivity: Identifies negative activities based on associated moods with categories, default negative activities.

Fetch user data:
getUserLocation: Fetches user location
fetchUserActivity: Fetches user activities for the previous week
fetchUserMoods: Fetches user moods for the previous week

Determine weather condition:
determineWeatherCondition: Determines weather condition based on temperature

Milestones:
Implemented fetchWeatherDataWithAPI and fetchWeatherDatafromDB to fetch weather data from API and database, respectively.
Developed getDayOfWeek, getTimeOfDay, and getMonthOfYear to extract relevant date information.
Created getUserLocation to fetch user location.
Implemented determineWeatherCondition to determine weather condition based on temperature.
Developed isNegativeActivity to identify negative activities based on associated moods and categories.
Created fetchUserActivity and fetchUserMoods to fetch user activities and moods for the previous week.

**What will be done in later PRs and not included here**
Integration of these helper functions with the main application logic.
Additional functionality to provide personalized recommendations/ goals based on user data

**Resources**
[https://developer.mozilla.org/en-US/docs/Web/API/Geolocation_API](url)
[https://www.weatherapi.com/docs/](url)

**Test Plan**
Unit tests for each helper function to ensure correctness.
Integration tests to verify the functions work together as expected
Edge cases tested include:
User has no activities
User has no moods
User location is not available

Note: Since this is a backend-only change, there are no UI-facing changes to test.

Video explanation:  [https://www.loom.com/share/4ca091c507b54aa4a7716bc07369e685](url)
Test plan video: [https://www.loom.com/share/717b5d1134f94b72ac71e68a4468ada3](url)

### CODE REVIEW

File: backend/ActivitiesAnalysis/fetchUserData.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+const { categorizeMood } = require('../DataAnalysis/analysis');
+
+// fetch the user's activities for the previous week
+const fetchUserActivity = async (userId) => {
+    try {
+        const activities = await prisma.userEntry.findMany({
+            where: {
+                userId: userId,
+                date: {
+                    gte: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000),
```
Comment (mulanzhao50) [inline backend/ActivitiesAnalysis/fetchUserData.js:15]: nit: If you are using this 7 * 24 * 60 * 60 * 1000 number multiple times, may be good to just turn it into a const variable 

like const DATE_RANGE = 7 * 24 * 60 * 60 * 1000 or something, and share it throughout this file.

**Review by mulanzhao50 — APPROVED** [review]
Thanks for incorporating the previous feedback to make your PRs more digestible! 

Additionally, I'm very impressed by this PR's structure and test plan. It's so great that you wrote unit tests to verify your backend code, and that you are iteratively thinking about the edge cases due to some struggles from TC#1. Love the proactive learning :)

File: backend/ActivitiesAnalysis/fetchUserData.js (hunk 1 above)
Comment (yourgirlmjayy) [reply backend/ActivitiesAnalysis/fetchUserData.js:15]: That sound's like a good idea. I'll implement that

**Comment by ghost** [comment]
+1 To what Mulan had said! The PR is much more digestible! Keep up the good work
